### DISCUSSION TITLE
Lofting through a non-planar curve

### DISCUSSION THREAD

**Discussion by MilesStarr** [gumyr/build123d#601] [Q&A] [answered]
Came here after running into limitations in CadQuery and much prefer the notation of the algebra system.  This is doing great, but I am running into difficulties getting a surface to generate from a closed set of 1d objects.  CadQuery can't do this either, but I am exporting step files to open in SolidWorks to do my actual lofting.  Having surfaces instead of individual 1D objects (SolidWorks imports each 1D object separately) is much nicer.  Is there any way to get a surface out of build123d for this complex geometry?  It has been wonderful for my flat profile lofts and would love to use build123d for my non-flat profiles.

![image](https://github.com/gumyr/build123d/assets/77509952/9a5789d7-4972-42e8-bb70-920984e63642)

I'm sure this is a bit obscure for a package not yet released to 1.0, so I understand if this isn't supported (yet?)

Keep up the great work!  I'll keep pushing the limits of the package and hopefully learn enough about the internal structure in the process that maybe I'll be able to contribute in the future.

**Comment by jdegenstein** [comment]
Can you clarify if you are facing an issue exporting geometry from build123d or creating geometry using build123d? I am not clear on what specifically you want to accomplish.

**Comment by gumyr** [comment, marked as answer]
Also not sure what you're looking for but one can create non-planar surfaces/faces from closed perimeter wires  - here is one example: https://build123d.readthedocs.io/en/latest/direct_api_reference.html#topology.Face.make_surface

The terminology of SolidWorks isn't exactly the same as buid123d so there might be a solution to your problem with a different term.

**Comment by MilesStarr** [comment]
Thank you, I didn't go into the API reference.  It worked exactly how I was expecting it to do, the make_face function died with this not being planar.  I'm in the middle of porting my code from CadQuery which failed to loft my previous work.  Perhaps I was having a similar issue with that package not going down to the API level.  Anyway since that wasn't getting me a functional solid, my workflow in CadQuery was:

- Generate parameters of cross sectional areas
- Generate 3D points along key locations of cross section by directly calculating where the need to be in space
- Generate 1D definitions of cross section
- Assemble Edges of cross sections
- Assemble Edges of loft guide curvatures passing through key points to ensure loft doesn't kink.
- (Failed to loft)
- Export edges to STEP
- Use the skeleton in the STEP file to make a loft.

Now I can use the @ and % operators to jump straight from my design parameters to 1D objects.  Now that I have a surface, I'll try lofting directly in build123d next week when I'm done porting my code over and will have multiple faces to loft through.  Worst case, exporting the surfaces will be much easier to work with compared to exporting 1D objects.

**Comment by gumyr** [comment]
There are two examples that might be helpful:
1) Lofting with many cross sections: https://build123d.readthedocs.io/en/latest/examples_1.html#multi-sketch-loft
2) Sweeping with many cross sections: https://build123d.readthedocs.io/en/latest/examples_1.html#handle

Keep in mind that 2D objects are required for both with can be created directly with `BuildSketch` / `Sketch` or from 1D objects created with `BuildLine` / `Curve` and converted to 2D objects with `make_face`. 

Note that in build123d / OCCT a loft has no path but a sweep does.

When setting up a multi-section sweep the `^` operator is very useful. `path ^ value` returns a `Location` along the `path` at the `value` (between 0.0 and 1.0) which contains both the position and orientation of the `path`.  This can be used to orient the cross sections appropriately to ensure they are always perpendicular to the path.
